keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,740 @@
+
+![alt text](./multi_numa.png "Multi-NUMA Partitioning")
+
+Figure above illustrates three groups of NUMA nodes. More details and the motivation behind the design of *multi-NUMA* strategy is present in [this](#how-to-enable-the-guaranteed-memory-allocation-over-many-numa-nodes) section.
+
+### User Stories
+
+#### Story 1 : High-performance packet processing with DPDK
+
+- Systems such as real-time trading system or 5G CNFs (User Plain Function, UPF) employ DPDK to guarantee low-latency packet processing. DPDK (*Data Plane Development Kit*) consists of a set of libraries designed for packet processing acceleration managed in software, in the userspace. DPDK requires dedicated resources (such as exclusive CPUs, hugepages, and DPDK-compatible Network Interface Card), and most importantly, the alignment of resources to the same NUMA node so as to prevent an unexpected performance degradation due to inter-node (between NUMA nodes) communication overhead. This creates the reason to guarantee the reservation of memory (and hugepages) and other computing resources (e.g. CPU cores and NICs) from the same NUMA node, for DPDK-oriented containers. One of key KPIs in 5G deployment is Ultra-Reliable and Low Latency Communications (URLLC) which is mission-critical. This accounts for why this proposal is also critical.
+
+#### Story 2 : Databases
+
+- Databases (e.g., Oracle, PostgreSQL, and MySQL) require relatively extensive amount of memory and hugepages to access massive volumes of data relatively efficiently. In order to reduce the latency caused by cross-NUMA memory access and sharing, all resources (CPU cores, memory, hugepages, and I/O devices) should be aligned to the same single NUMA node. This will serve the stability and performance enhancement significantly.
+
+```
+<<[UNRESOLVED]>>
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+TBD
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+TBD
+
+<<[/UNRESOLVED]>>
+```
+
+## Design Details 
+
+### How to enable the guaranteed memory allocation over many NUMA nodes?
+
+Our initial concept concerning the guaranteed memory allocation had a limitation, and for each container, the container's memory consumption was limited to exactly one source, i.e., a single NUMA node. That is why Memory Menager was planned to feature only the single-numa-node policy, but it was in the past, and it is not true anymore.  
+
```

Comment:
I'm not sure we need to mention previous ideas in this document. It makes understanding harder in my opinion.

Reply:
Generally, I think, it's good practice to provide background information even if it is lengthy. It helps to understand the motivation which led to the present proposal.
Nevertheless, I followed your suggestion, and I commented out the introductory passages.